### PR TITLE
Add support for restore cart

### LINKED ISSUE #87: Add support for the restore-cart link

### PR DESCRIPTION
**Pull request by chenliangde**
Closes #87

### CODE REVIEW

File: Helper/Url.php [hunk 1]
```diff
@@ -233,11 +306,82 @@ public function getPreviewUrlFront(Store $store)
...
+    /**
+     * Gets the absolute URL to the current store view cart page.
+     *
+     * @param Store $store the store to get the url for.
+     * @param string $currentUrl restore cart url
+     * @return string the url.
+     */
+    public function getUrlCart(Store $store, $currentUrl)
+    {
+        // @codingStandardsIgnoreLine
```
Comment (mridang) [inline Helper/Url.php]: This seems wrong.

File: Helper/Data.php [hunk 2]
```diff
@@ -117,6 +117,11 @@ class Data extends AbstractHelper
      */
     const PLATFORM_NAME = 'Magento';
 
+    /**
+     * @var string the algorithm to use for hashing visitor id.
+     */
+    const VISITOR_HASH_ALGO = 'sha256';
```
Comment (mridang) [inline Helper/Data.php]: Remove this and move the hashing to `Magento\Framework\Encryption\Encryptor`

File: Model/Customer.php [hunk 3]
```diff
@@ -138,4 +138,25 @@ public function _construct()
...
+    public function getRestoreCartHash()
+    {
+        return $this->getData(self::RESTORE_CART_HASH);
+    }
+
+    /**
+     * Set restore cart hash
+     *
+     * @param string $restoreCartHash
+     * @return self
```
Comment (mridang) [inline Model/Customer.php]: Does this really return `self`?

File: CustomerData/CartTagging.php [hunk 4]
```diff
@@ -41,11 +41,14 @@
...
 use Magento\Framework\Stdlib\DateTime\DateTime;
 use Nosto\Object\Cart\LineItem;
+use Nosto\Tagging\Helper\Data;
 use Nosto\Tagging\Helper\Scope as NostoHelperScope;
+use Nosto\Tagging\Helper\Scope;
 use Nosto\Tagging\Model\Cart\Builder as NostoCartBuilder;
 use Nosto\Tagging\Model\Customer;
 use Nosto\Tagging\Model\Customer as NostoCustomer;
 use Nosto\Tagging\Model\CustomerFactory as NostoCustomerFactory;
+use Nosto\Tagging\Helper\Url;
```
Comment (mridang) [inline CustomerData/CartTagging.php]: Correct the alias. I think we import it as `NostoHelperUrl` or something like that.

File: CustomerData/CartTagging.php [hunk 5]
```diff
@@ -41,11 +41,14 @@
 use Magento\Framework\Stdlib\CookieManagerInterface;
 use Magento\Framework\Stdlib\DateTime\DateTime;
 use Nosto\Object\Cart\LineItem;
+use Nosto\Tagging\Helper\Data;
 use Nosto\Tagging\Helper\Scope as NostoHelperScope;
+use Nosto\Tagging\Helper\Scope;
```
Comment (mridang) [inline CustomerData/CartTagging.php]: This is redundant. See above.

File: Helper/Url.php [hunk 6]
```diff
@@ -233,11 +306,82 @@ public function getPreviewUrlFront(Store $store)
...
+    }
+
+    /**
+     * Returns restore cart url
+     *
+     * @param string $hash
+     * @param Store $store
+     * @return string
+     */
+    public function generateRestoreCartUrl($hash, Store $store)
```
Comment (mridang) [inline Helper/Url.php]: I would move this method into it's own class `Nosto\Tagging\Model\Cart\Restore\Builder.php`

File: Helper/Url.php (hunk 1 above)
Comment (chenliangde) [inline Helper/Url.php]: It suppress phpcs check for the next line

File: CustomerData/CartTagging.php [hunk 7]
```diff
@@ -122,7 +129,12 @@ public function getSectionData()
         }
 
         if ($data["itemCount"] > 0) {
-            $this->updateNostoId();
+            $nostoCustomer = $this->updateNostoId();
+            if ($nostoCustomer && $nostoCustomer->getRestoreCartHash()) {
```
Comment (mridang) [inline CustomerData/CartTagging.php]: Maybe we can wrap this in a feature-flag?

File: CustomerData/CartTagging.php [hunk 8]
```diff
@@ -122,7 +130,9 @@ public function getSectionData()
         }
 
         if ($data["itemCount"] > 0) {
-            $this->updateNostoId();
+            $store = $this->scopeHelper->getStore();
+            $data['restore_cart_url'] = $this->nostoRestoreCartUrlBuilder
```
Comment (hannunosto) [inline CustomerData/CartTagging.php]: I would add `$data['restore_cart_url'] = null` also to here https://github.com/Nosto/nosto-magento2/blob/develop/CustomerData/CartTagging.php#L98 to have the same data structure regardless if the cart has items or not

File: Model/Cart/Restore/Builder.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+     * @param LoggerInterface $logger
+     * @param CookieManagerInterface $cookieManager
+     * @param NostoCustomerFactory $nostoCustomerFactory
+     * @param NostoHelperUrl $urlHelper
+     * @param DateTime $date
+     */
+    public function __construct(
+        LoggerInterface $logger,
+        CookieManagerInterface $cookieManager,
+        EncryptorInterface $encryptor,
```
Comment (hannunosto) [inline Model/Cart/Restore/Builder.php]: This is missing from doc block

Comment (chenliangde) [inline Model/Cart/Restore/Builder.php]: Fixed

File: CustomerData/CartTagging.php [hunk 10]
```diff
@@ -40,52 +40,60 @@
...
-        NostoCustomerFactory $nostoCustomerFactory
+        NostoCustomerFactory $nostoCustomerFactory,
+        NostoRestoreCartUrlBuilder $nostoRestoreCartUrlBuilder
     ) {
         $this->cartHelper = $cartHelper;
-        $this->nostoCartBuilder = $nostoCartBuilder;
         $this->logger = $logger;
         $this->date = $date;
         $this->cookieManager = $cookieManager;
+        $this->scopeHelper = $scopeHelper;
```
Comment (hannunosto) [inline CustomerData/CartTagging.php]: Why was this renamed? Now it's not consistent with other Nosto classes
